src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/aggregators/PersistableSlidingWindow.java

```diff
@@ -0,0 +1,103 @@
+import com.fasterxml.jackson.databind.ObjectMapper;
+import java.io.BufferedWriter;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Iterator;
+import java.util.concurrent.TimeUnit;
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.io.LineIterator;
+import org.apache.commons.lang3.RandomStringUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * PersistableSlidingWindow is a SlidingWindow which can have its data written to and read from disk
+ */
+public class PersistableSlidingWindow extends SlidingWindow<SlidingWindowData> {
+  private static final Logger LOG = LogManager.getLogger(PersistableSlidingWindow.class);
+  private static final ObjectMapper objectMapper = new ObjectMapper();
+
+  protected static final String SEPARATOR = "\n";
+
+  private Path pathToFile;
+
+  public PersistableSlidingWindow(int slidingWindowSize,
+                                  TimeUnit timeUnit,
+                                  Path filePath) {
+    super(slidingWindowSize, timeUnit);
+    this.pathToFile = filePath;
+    try {
+      if (Files.exists(pathToFile)) {
+        load(pathToFile);
+      } else {
+        Files.createFile(pathToFile);
+      }
+    } catch (IOException e) {
+      LOG.error("Couldn't create file {} to perform young generation tuning", pathToFile, e);
+      throw new IllegalArgumentException("Couldn't create or read a file at " + filePath);
+    }
+  }
+
+  /**
+   * Loads the SlidingWindowData contained in the given path into this PersistableSlidingWindow
+   * @param path The path to the file containing the SlidingWindow data
+   * @throws IOException If there is an error reading the file
+   */
+  protected void load(Path path) throws IOException {
+    LineIterator it = FileUtils.lineIterator(path.toFile(), "UTF-8");
+    try {
+      while (it.hasNext()) {
+        String line = it.nextLine();
+        SlidingWindowData data = objectMapper.readValue(line, SlidingWindowData.class);
+        next(data);
+      }
+    } finally {
+      LineIterator.closeQuietly(it);
+    }
+  }
+
+  /**
+   * Writes the contents of this SlidingWindow into the path provided during construction
+   *
+   * <p>This write occurs in 2 stages, it writes to a temporary file, then replaces the actual data
+   * file with the contents of the temporary file
+   *
+   * @throws IOException If there is a CRUD error with the files involved
+   */
+  protected void write() throws IOException {
+    String tmpFile = pathToFile.toString() + RandomStringUtils.randomAlphanumeric(32);
+    Path tmpPath = Paths.get(tmpFile);
+    Files.createFile(Paths.get(tmpFile));
+    BufferedWriter writer = new BufferedWriter(new FileWriter(tmpFile, false));
+    Iterator<SlidingWindowData> it = windowDeque.descendingIterator();
```

Comment:
any reason why we're writing in the reverse sequential order? Won't this bring back the sliding window in the reverse order after a load?

Reply:
This is so that things are in the correct order. Look at how SlidingWindow stores data in a LinkedList on calls to next(). Also see PersistableSlidingWindowTest#testNext(). I added an extra case to this function for extra sanity as well.

Reply:
ha! We offerFirst() and pollLast(), it makes sense now.